src/ipv6.ts

```diff
@@ -121,19 +121,108 @@ export function isInSubnet(address: string, subnetOrSubnets: string | string[]):
   return true;
 }
 
+/**
+ * Create a function to test if a given IPv6 address is contained in the specified subnet.
+ * @param subnet the IPv6 CIDR to test (or an array of them)
+ * @throws if the subnet(s) are not valid IP addresses, or the CIDR prefix lengths
+ *  are not valid
+ */
+export function createChecker(
+  subnetOrSubnets: string | string[]
+): (address: string) => boolean {
+  if (Array.isArray(subnetOrSubnets)) {
+    const checks = subnetOrSubnets.map(subnet => createSegmentChecker(subnet));
+    return address => {
+      const segments = getIpv6Segments(address);
+      return checks.some(check => check(segments));
+    };
+  }
+  const check = createSegmentChecker(subnetOrSubnets);
+  return address => {
+    const segments = getIpv6Segments(address);
+    return check(segments);
+  };
+}
+
+// This creates the last function that works on the most deconstructed data
+function createSegmentChecker(subnet: string): (segments: string[]) => boolean {
+  const [subnetAddress, prefixLengthString] = subnet.split('/');
+  const prefixLength = parseInt(prefixLengthString, 10);
+
+  if (!subnetAddress || !Number.isInteger(prefixLength)) {
+    throw new Error(`not a valid IPv6 CIDR subnet: ${subnet}`);
+  }
+
+  if (prefixLength < 0 || prefixLength > 128) {
+    throw new Error(`not a valid IPv6 prefix length: ${prefixLength} (from ${subnet})`);
+  }
+
+  // the next line throws if the address is not a valid IPv6 addresse
+  const subnetSegments = getIpv6Segments(subnetAddress);
+
+  return addressSegments => {
+    for (let i = 0; i < 8; ++i) {
+      const bitCount = Math.min(prefixLength - i * 16, 16);
+
+      if (bitCount <= 0) {
+        break;
+      }
+
+      const subnetPrefix =
+        ((subnetSegments[i] && parseInt(subnetSegments[i], 16)) || 0) >> (16 - bitCount);
+
+      const addressPrefix =
+        ((addressSegments[i] && parseInt(addressSegments[i], 16)) || 0) >>
+        (16 - bitCount);
+
+      if (subnetPrefix !== addressPrefix) {
+        return false;
+      }
+    }
+
+    return true;
+  };
+}
+
+// cache these special subnet checkers
+const specialNetsCache: Record<string, (address: string) => boolean> = {};
+function getSpecialChecker(
```

Comment:
I don’t think this is needed. Each method below can manage its own cache entry. While that adds a bit of code to `isPrivate`, `isLocalhost` etc., it avoids having to hardcode the `kind` here. Better to hardcode the word “private” just once right in the `isPrivate` method. (Same for the equivalent function in `ipv4.ts`.)